### PR TITLE
Add docstrings and comments

### PR DESCRIPTION
**Pull request by LuEdRaMo**

### CODE REVIEW

**Review by lbenet — COMMENTED** [review]
Thanks a lot for your impressive work! This is truly an important step forward to build a proper documentation.
I've left several comments, mostly related with completing a reference, typos, etc. Others are not relevant for this PR, but maybe interesting to consider in the future, so I suggest you don't "resolve" them to keep them in mind. They are related to possible microperformance  improvements, and/or type-stability issues. 
All in all, LGTM! So I suggest you go through the comments, and once you have addressed them (except for those that involve other PRs), we go ahead and merge this PR. I don't think tagging a new (patch) version is important now.
Truly, thanks a lot!

**Comment by lbenet** [comment]
cc @PerezHz

File: src/PlanetaryEphemeris.jl [hunk 1]
```diff
@@ -29,10 +29,18 @@ using Quadmath
 
 import Base.reverse
 
-# integration parameters
+# Integration parameters
 const order = 30
 const abstol = 1.0E-20
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/PlanetaryEphemeris.jl:34]: I've moved both constants to `constants.jl`.

File: src/constants.jl [hunk 2]
```diff
@@ -1,8 +1,9 @@
...
-const ea = 4 #Earth's index
-const mo = 5 #Moon's index
+const su = 1 # Sun's index
+const ea = 4 # Earth's index
+const mo = 5 # Moon's index
 
-# vector of G*m values in au^3/day^2
+# Mass parameters μ = G*m (in au^3/day^2) of major bodies (Sun + Planets + Moon)
+# See Table 8 in page 49 of https://ui.adsabs.harvard.edu/abs/2014IPNPR.196C...1F%2F/abstract
 GM1 = 0.491248045036476000E-10 # Mercury
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/constants.jl:7]: Done.

File: src/constants.jl [hunk 3]
```diff
@@ -500,69 +535,91 @@ const C65M =  1.1737698784460500E-09
...
-
-#diagonal elements of undistorted lunar mantle moment of inertia
+SM[3,1:3] = S3M    # n = 3
+SM[4,1:4] = S4M    # n = 4
+SM[5,1:5] = S5M    # n = 5
+SM[6,1:6] = S6M    # n = 6
+
+# Extended body parameters for the Moon
+# See Table 11 in page 51 of https://ui.adsabs.harvard.edu/abs/2014IPNPR.196C...1F%2F/abstract
+const R_moon = RM/au                         # Lunar radius in au
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/constants.jl:557]: There were two sections in `constants.jl` named `Extended body parameters for the Moon`. I've joined them and change all occurrences of `RM/au` for `R_moon`.

File: src/jpl-de-430-431-earth-orientation-model.jl [hunk 4]
```diff
@@ -95,6 +223,26 @@ end
...
+```
+where ``\alpha`` is an angle in radians. 
+
+See equation (11) in page 9 of https://ui.adsabs.harvard.edu/abs/2014IPNPR.196C...1F%2F/abstract.
+
+See also [`Ry`](@ref) and [`Rz`](@ref).
+"""
 function Rx(alpha::T) where {T<:Number}
     res = Array{T}(undef, 3, 3)
     res[1, 1] = one(alpha)
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/jpl-de-430-431-earth-orientation-model.jl:248]: Done.

File: src/jpl-de-430-431-earth-orientation-model.jl [hunk 5]
```diff
@@ -109,8 +257,31 @@ function Rx(alpha::T) where {T<:Number}
...
+``\textbf{Note:}`` Lieske et al. (1977, 1979) introduces the convention that this matrix 
+has opposite signs outside the sine terms (with respect to `Rx`, `Rz`). See equation (4) 
+in page 3 of https://ui.adsabs.harvard.edu/abs/1977A%26A....58....1L/abstract and equation 
+(3) in page 283 of https://ui.adsabs.harvard.edu/abs/1979A%26A....73..282L/abstract.
+
+See also [`Rx`](@ref) and [`Rz`](@ref).
+"""
 function Ry(alpha::T) where {T<:Number}
     res = Array{T}(undef, 3, 3)
     res[1, 1] = cos(alpha)
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/jpl-de-430-431-earth-orientation-model.jl:287]: Done.

File: src/jpl-de-430-431-earth-orientation-model.jl [hunk 6]
```diff
@@ -125,6 +296,26 @@ function Ry(alpha::T) where {T<:Number}
...
+```
+where ``\alpha`` is an angle in radians. 
+
+See equation (13) in page 9 of https://ui.adsabs.harvard.edu/abs/2014IPNPR.196C...1F%2F/abstract.
+
+See also [`Rx`](@ref) and [`Ry`](@ref).
+"""
 function Rz(alpha::T) where {T<:Number}
     res = Array{T}(undef, 3, 3)
     res[1, 1] = cos(alpha)
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/jpl-de-430-431-earth-orientation-model.jl:321]: Done. Perhaps a function that returns the three rotation matrices can save even more memory by computing sines/cosines only once.

File: src/constants.jl [hunk 7]
```diff
@@ -500,69 +535,91 @@ const C65M =  1.1737698784460500E-09
...
-const f_c = 2.4623904789198150E-04 #oblateness of core
-const k_ν_div_C_T = 1.6365616533709530E-08 # Friction coefficient between core and mantle, radian/day
-
-const J2_M_und = 2.0321568464952570E-04 # undistorted lunar 2nd zonal harmonic coefficient
-
-#diagonal elements of undistorted lunar mantle moment of inertia
+SM[3,1:3] = S3M    # n = 3
+SM[4,1:4] = S4M    # n = 4
+SM[5,1:5] = S5M    # n = 5
+SM[6,1:6] = S6M    # n = 6
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/constants.jl:553]: Done.

File: src/initial_conditions.jl [hunk 8]
```diff
@@ -1,51 +1,75 @@
+@doc raw"""
+    initialcond(N, jd0::Float64=datetime2julian(DateTime(1969,6,28,0,0,0)))
+
+Returns a vector with the initial conditions (`3N` positions [au] + `3N` velocities [au/day] + 
+3 lunar mantle Euler angles [rad] + 3 mantle angular velocity [rad/day] +
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/initial_conditions.jl:5]: Corrected.

File: src/dynamical_model.jl [hunk 9]
```diff
@@ -1,16 +1,20 @@
 # Solar System ( JPL DE430/431) dynamical model
 # Bodies considered in the model are: the Sun, the eight planets, the Moon and
-# the 343 main-belt asteroids included in the JPL DE 430 ephemeris
+# the 343 main-belt asteroids included in the JPL DE430 ephemeris
 # effects considered are:
 # - post-Newtonian point-mass accelerations between all bodies,
 # - figure-effects (oblateness) of the Earth (J2 and J3)
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:6]: Done.

File: src/dynamical_model.jl [hunk 10]
```diff
@@ -1,16 +1,20 @@
...
 # - Kinematic model for the precession and nutation of the Earth's orientation (IAU 1976/1980 Earth orientation model)
 # - Kinematic model for the Moons's orientation (Seidelmann et al., 2006)
-# - tidal secular acceleration of Moon due to rides raised on Earth by both the Moon and the Sun
+# - Tidal secular acceleration of Moon due to rides raised on Earth by both the Moon and the Sun
 
-# auxiliary function: order-preserving differentiate
+@doc """
+    ordpres_differentiate(a::Taylor1)
+
+Auxiliary function: order-preserving differentiate.
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:16]: Done.

File: src/dynamical_model.jl [hunk 11]
```diff
@@ -19,97 +23,253 @@ function ordpres_differentiate(a::Taylor1)
...
+    local I_m_t = (ITM_und-I_c).*one_t           # Undistorted moment of inertia of the mantle, see equation (40)
+    local dI_m_t = ordpres_differentiate.(I_m_t) # Time-derivative of lunar mantle I at time t-τ_M
+    local inv_I_m_t = inv(I_m_t)                 # Inverse of lunar mantle I matrix at time t-τ_M
+    local I_c_t = I_c.*one_t                     # Lunar core I matrix, see equation (39)
+    local inv_I_c_t = inv(I_c_t)                 # Inverse of lunar core I matrix
+    local I_M_t = I_m_t+I_c_t                    # Total I matrix (mantle + core)
+
+    # Parameters related to speed of light, c
+    local c_p2 = 29979.063823897606              # c^2 = 29979.063823897606 au^2/d^2
+    local c_m2 = 3.3356611996764786e-5           # c^-2 = 3.3356611996764786e-5 d^2/au^2
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:162]: The three functions in `dynamical_model.jl` use the same `c_p2` and `c_m2` as local variables. I've moved both to `constants.jl`.

File: src/dynamical_model.jl [hunk 12]
```diff
@@ -19,97 +23,253 @@ function ordpres_differentiate(a::Taylor1)
...
+
+    #=
+    Point-mass accelerations 
+    See equation (35) in page 7 of https://ui.adsabs.harvard.edu/abs/1971mfdo.book.....M/abstract
+    =#
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:167]: I've added a note to the three functions in `dynamical_model.jl`.

File: src/dynamical_model.jl [hunk 13]
```diff
@@ -19,97 +23,253 @@ function ordpres_differentiate(a::Taylor1)
...
-    newton_acc_Y = Array{S}(undef, N, N)
-    newton_acc_Z = Array{S}(undef, N, N)
-
-    v2 = Array{S}(undef, N)
-    _2v2 = Array{S}(undef, N, N)
-    vi_dot_vj = Array{S}(undef, N, N)
-    pn2 = Array{S}(undef, N, N)
+    # Newtonian potential of N bodies 
+    # \sum_{i\neq l} \frac{\mu_i}{r_{il}} or
+    # \sum_{j\neq k} \frac{\mu_j}{r_{jk}}
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:208]: Done.

File: src/dynamical_model.jl [hunk 14]
```diff
@@ -19,97 +23,253 @@ function ordpres_differentiate(a::Taylor1)
...
-
-    pntempX = Array{S}(undef, N)
-    pntempY = Array{S}(undef, N)
-    pntempZ = Array{S}(undef, N)
+    # Full third term of equation (35)
+    pNX_t_pn3 = Array{S}(undef, N, N)      # X-axis component
+    pNY_t_pn3 = Array{S}(undef, N, N)      # Y-axis component
+    pNZ_t_pn3 = Array{S}(undef, N, N)      # Z-axis component
+
+    # First term
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:236]: Done.

File: src/dynamical_model.jl [hunk 15]
```diff
@@ -218,126 +421,166 @@ end
...
-    local J2S_t = JSEM[su,2]*one_t
-    J2_t = Array{S}(undef, 5)
-    J2_t[su] = J2S_t
-    J2_t[ea] = J2E_t
-    # lunar torques: overall numerical factor in Eq. (44) of Folkner et al. (2014)
+    local RE_au = (RE/au)                              # Earth's radius in au
+    # Second zonal harmonic coefficient
+    # See Table 10 in page 50 of https://ui.adsabs.harvard.edu/abs/2014IPNPR.196C...1F%2F/abstract
+    local J2E_t = (J2E + J2EDOT*(dsj2k/yr))*(RE_au^2)  # Earth (considering a linear change in time with rate J2EDOT)
+    local J2S_t = JSEM[su,2]*one_t                     # Sun (static)
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:450]: `RE_au` is indeed a constant, so I've moved it to `constants.jl`. However, `J2E_t` does change with time, so I will leave it there.

File: src/osculating.jl [hunk 16]
```diff
@@ -1,85 +1,100 @@
...
 
-Returns semimajor axis `a`, eccentricity `e` and inclination `I` for the two-body
+Returns semimajor axis `a`, eccentricity `e` and inclination `inc` for the two-body
 problem defined by the relative position `(x,y,z)` and velocity `(u,v,w)`vectors
-between two bodies with masses `m1` and `m2`."""
+between two bodies with masses `m1` and `m2`.
+
+See also [`semimajoraxis`](@ref), [`eccentricity`](@ref) and [`inclination`](@ref).
+"""
 function aei(x, y, z, u, v, w, m1, m2)
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/osculating.jl:50]: `eccentricity` computes `semimajoraxis`, so `aei` avoids repeating a couple of calculations. Something similar happens with `inclination`.

File: src/dynamical_model.jl [hunk 17]
```diff
@@ -218,126 +421,166 @@ end
...
+    local fact5_jsem = [(n+2) for n in 1:maximum(n1SEM)]     # n + 2
+    # Numerical factors for recursion relations of Associated Legendre polynomials
+    # See equations (180)-(183) in pages 33-34 of https://ui.adsabs.harvard.edu/abs/1971mfdo.book.....M/abstract
+    local lnm1 = [(2n-1)/(n-m) for n in 1:6, m in 1:6]       # (2n - 1) / (n - m)
+    local lnm2 = [-(n+m-1)/(n-m) for n in 1:6, m in 1:6]     # -(n + m - 1) / (n - m)
+    local lnm3 = [-n for n in 1:6]                           # -n 
+    local lnm4 = [n+m for n in 1:6, m in 1:6]                # (n + m)
+    local lnm5 = [2n-1 for n in 1:6]                         # (2n - 1)
+    local lnm6 = [-(n+1) for n in 1:6]                       # -(n + 1)
+    local lnm7 = [m for m in 1:6]                            # m
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:444]: I've moved all `facti_jsem` and `lnmi` to `constants.jl`.

File: src/osculating.jl [hunk 18]
```diff
@@ -1,85 +1,100 @@
...
+    ae(x, y, z, u, v, w)
 
 Returns semimajor axis `a` and eccentricity `e` for the two-body problem defined
 by the relative position `(x,y,z)` and velocity `(u,v,w)` vectors between two
-bodies with masses `m1` and `m2`."""
+bodies with masses `m1` and `m2`.
+
+See also [`semimajoraxis`](@ref) and [`eccentricity`](@ref).
+"""
 function ae(x, y, z, u, v, w, m1, m2)
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/osculating.jl:75]: Same as with `aei`, `ae` avoids repeating a couple of calculations.

File: src/dynamical_model.jl [hunk 19]
```diff
@@ -19,97 +23,253 @@ function ordpres_differentiate(a::Taylor1)
...
 end
 
+@doc raw"""
+    NBP_pN_A_J23E_J23M_J2S!(dq, q, params, t)
+
+Solar System (JPL DE430/431) dynamical model. Bodies considered in the model are: the Sun, 
+the eight planets, the Moon and the 343 main-belt asteroids included in the JPL DE430 
+ephemeris. Effects considered are:
+
+- Post-Newtonian point-mass accelerations between all bodies: see equation (35) in page 7 of https://ui.adsabs.harvard.edu/abs/1971mfdo.book.....M/abstract
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:33]: I've checked the equations again and only modify the indices in the post-newtonian accelerations equation.

File: src/dynamical_model.jl [hunk 20]
```diff
@@ -685,130 +1017,186 @@ end
...
 
+@doc raw"""
+    NBP_pN_A_J23E_J23M_J2S_threads!(dq, q, params, t)
+
+Threaded version of `NBP_pN_A_J23E_J23M_J2S!`.
+
+See also [`NBP_pN_A_J23E_J23M_J2S!`](@ref).
+""" NBP_pN_A_J23E_J23M_J2S_threads!
+
 @taylorize function NBP_pN_A_J23E_J23M_J2S_threads!(dq, q, params, t)
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:1067]: I've applied all the relevant comments to the three functions in `dynamical_model.jl`.

File: src/dynamical_model.jl [hunk 21]
```diff
@@ -1418,114 +1982,200 @@ end
...
+the position vectors ``\mathbf{r}`` and ``\mathbf{r}_j^*`` with ``j = 0, 1, 2`` are
+expressed in cylindrical coordinates with the ``Z`` axis perpendicular to the Earth’s equator,
+so that ``\mathbf{r} = \mathbf{\rho} + \mathbf{z}`` and the time-delayed position of the tide-raising
+body is given by ``\mathbf{r}_j^* = \mathbf{\rho}_j^* + \mathbf{z}_j^*``; and ``\mathbf{a}_{M,tide}`` is
+the acceleration of the Moon with respect to Earth, for each tide-raising body.
+
+See also [`NBP_pN_A_J23E_J23M_J2S!`](@ref) and [`NBP_pN_A_J23E_J23M_J2S_threads!`](@ref).
+""" DE430!
+
 @taylorize function DE430!(dq, q, params, t)
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/dynamical_model.jl:2021]: Done.

File: src/propagation.jl [hunk 22]
```diff
@@ -1,42 +1,89 @@
+
+@doc raw"""
+    selecteph2jld(sseph::TaylorInterpolant, bodyind::AbstractVector{Int}, tspan::Number, N::Int)
+
+Saves the ephemeris, contained in `sseph`, of the bodies with indexes `bodyind`, in a jld file
+named as follows
+
+    sseph + number of asteroids in sseph + ast + number of asteroids to be saved in file 
+    + p (forward integration) or n (backward integration) + y_et.jld
```
[the remark this answers is not in the document]
Comment (LuEdRaMo) [reply src/propagation.jl]: Done.

**Comment by LuEdRaMo** [comment]
In the last commit I've addressed most of the comments. Those unresolved may be the subject of another PR.

**Comment by LuEdRaMo** [comment]
I've run `scripts/integrate_ephemeris.jl` and the results of the integration do not change with the commits in this PR.

**Comment by lbenet** [comment]
> I've run scripts/integrate_ephemeris.jl and the results of the integration do not change with the commits in this PR.

Excellent! So let's go ahead and merge this PR. 

Thanks truly a lot for the enormous work put in commenting everything!

**Comment by PerezHz** [comment]
Amazing job @LuEdRaMo! Many thanks for all the work you put here!

**Pull request merged**